app/grandchallenge/algorithms/views.py

```diff
@@ -1013,3 +1026,56 @@ def get_success_message(self, cleaned_data):
 
     def get_success_url(self):
         return self.algorithm.get_absolute_url() + "#models"
+
+
+class AlgorithmImageTemplate(ObjectPermissionRequiredMixin, DetailView):
+    model = Algorithm
+    permission_required = "algorithms.change_algorithm"
+    raise_exception = True
+    queryset = Algorithm.objects.prefetch_related(
+        "inputs__json_kind_example",
+        "outputs__json_kind_example",
+    )
+
+    def get(self, *_, **__):
+        algorithm = self.get_object()
+
+        forge_context = get_forge_algorithm_template_context(algorithm)
+
+        with tempfile.TemporaryDirectory() as temp_dir:
+            algorithm_template_path = generate_algorithm_template(
+                context=forge_context,
+                output_path=Path(temp_dir),
+            )
+
+            return FileResponse(
+                streaming_content=self.zip_memory_buffer(
+                    source_path=algorithm_template_path,
+                    allowed_base_path=Path(temp_dir),
+                ),
+                as_attachment=True,
+                filename=f"{algorithm.slug}-template.zip",
+                content_type="application/zip",
+            )
+
+    @staticmethod
+    def zip_memory_buffer(*, source_path, allowed_base_path):
+        buffer = io.BytesIO()
+        with ZipFile(buffer, "w") as zipf:
+            for foldername, _, filenames in os.walk(
+                source_path, followlinks=True
+            ):
+                for filename in filenames:
+                    file_path = Path(foldername) / filename
+                    file_path = file_path.resolve()
+
+                    if allowed_base_path not in file_path.parents:
```

Comment:
I would be more comfortable with using Django's implementation rather than implementing this ourselves.

Reply:
Took a bit of head scratching but I decided to sidestep the issue all together: it now packs all the content of the `temp_dir`. This entails an extra subdirectory in the ZIP but simplifies the trust issue.

I kep the checks on the resolved path to catch relative symlinks which might be a potential vector for unintended additions to the ZIP.